src/ioctls/kvm_vec.rs

```diff
@@ -0,0 +1,405 @@
+        (size_in_bytes + size_of::<T>() - 1) / size_of::<T>()
+    }
+
+    /// Constructs a new KvmVec<T> that contains `num_elements` empty elements of type `KvmArray::Entry`
+    ///
+    pub fn new(num_elements: usize) -> KvmVec<T> {
+        let required_kvm_array_capacity = KvmVec::<T>::num_elements_to_kvm_array_len(num_elements);
+
+        let mut kvm_array = Vec::with_capacity(required_kvm_array_capacity);
+        for _ in 0..required_kvm_array_capacity {
+            kvm_array.push(T::default())
+        }
+        kvm_array[0].set_len(num_elements);
+
+        KvmVec {
+            kvm_array,
+            len: num_elements,
+            capacity: num_elements,
+        }
+    }
+
+    /// Returns a reference to the actual KVM structure instance.
+    ///
+    pub fn as_original_struct(&self) -> &T {
+        &self.kvm_array[0]
+    }
+
+    /// Returns a mut reference to the actual KVM structure instance.
+    ///
+    pub fn as_mut_original_struct(&mut self) -> &mut T {
+        &mut self.kvm_array[0]
+    }
+
+    /// Get a pointer to the KVM struct so it can be passed to the kernel.
+    ///
+    pub fn as_ptr(&self) -> *const T {
+        self.as_original_struct()
+    }
+
+    /// Get a mutable pointer to the KVM struct so it can be passed to the kernel.
+    ///
+    pub fn as_mut_ptr(&mut self) -> *mut T {
+        self.as_mut_original_struct()
+    }
+
+    /// Returns a mut `Vec<KvmArray::Entry>` that contains all the elements.
+    /// It is important to call `mem::forget` after using this vector. Otherwise rust will destroy it.
+    ///
+    unsafe fn as_vec(&mut self) -> Vec<T::Entry> {
+        let entries_ptr = self.as_mut_original_struct().entries_mut().as_mut_ptr();
+        Vec::from_raw_parts(entries_ptr, self.len, self.capacity as usize)
+    }
+
+    /// Get the mutable elements slice so they can be modified before passing to the VCPU.
+    ///
+    pub fn as_entries_slice(&self) -> &[T::Entry] {
+        let len = self.as_original_struct().len();
+        unsafe { self.as_original_struct().entries().as_slice(len as usize) }
+    }
+
+    /// Get the mutable elements slice so they can be modified before passing to the VCPU.
+    ///
+    pub fn as_mut_entries_slice(&mut self) -> &mut [T::Entry] {
+        let len = self.as_original_struct().len();
+        unsafe {
+            self.as_mut_original_struct()
+                .entries_mut()
+                .as_mut_slice(len as usize)
+        }
+    }
+
+    /// Reserves capacity for at least `additional` more `KvmArray::Entry` elements.
+    ///
+    fn reserve(&mut self, additional: usize) {
+        let current_kvm_array_len = self.kvm_array.len();
+        let required_kvm_array_len =
+            KvmVec::<T>::num_elements_to_kvm_array_len(self.len + additional);
+        let additional_kvm_array_len = required_kvm_array_len - current_kvm_array_len;
+
+        self.kvm_array.reserve(additional_kvm_array_len);
+        self.capacity = self.len + additional;
```

Comment:
I might be confused, but the `reserve()` is done on logic that's based on current `len`, but here you're also increasing capacity with the same amount.

Isn't this broken in the case where `len` is smaller than `capacity`?

Reply:
You're right. I missed this case. Fixed.